src/java.base/share/classes/jdk/internal/foreign/abi/SharedUtils.java

```diff
@@ -162,12 +162,20 @@ public static MethodHandle adaptDowncallForIMR(MethodHandle handle, FunctionDesc
      * @param target the target handle to adapt
      * @return the adapted handle
      */
-    private static MethodHandle adaptUpcallForIMR(MethodHandle target, boolean dropReturn) {
+    private static MethodHandle adaptUpcallForIMR(MethodHandle target, boolean dropReturn, boolean hasCaptureCallState) {
         if (target.type().returnType() != MemorySegment.class)
             throw new IllegalArgumentException("Must return MemorySegment for IMR");
 
         target = collectArguments(MH_BUFFER_COPY, 1, target); // (MemorySegment, ...) MemorySegment
 
+        if (hasCaptureCallState) {
+            assert target.type().parameterType(1) == MemorySegment.class;
+            // IMR is an ABI figment, so the capture state segment which
+            // is inserted by CallingSequenceBuilder as leading parameter
+            // should appear before the IMR segment parameter
+            target = swapArguments(target, 0, 1);
+        }
```

Comment:
e.g. 
```
target = (MemorySegment, MemorySegment...) MemorySegment
          ^ IMR segment  ^ CCS segtment
```

These should be swapped, since the binding recipe for the capture state segment comes before the binding recipe for the IMR segment, since the former is inserted as a leaded parameter after a particular CallArranger inserts the IMR segment parameter).